doc/internals/file-system.md

```diff
@@ -0,0 +1,206 @@
+(success).
+
+All file system host classes include file-host-mount-mixin and when a
+file system host is created, the mount-args slot is set to a value
+that depends on both the host type and the file system associated with
+the host. For example, for a FAT32 host, the mount-args slot would be
+set to partition UUID, and for a NFS host, the mount-args slot might
+be set to the url of the NFS file system, e.g.,
+"nfs://nfs.lispm.net/export/home/tom".
+
+During the boot process, after the disks and partitions are
+enumerated, but before any file system references occur, mount is
+called for each of the host objects registered with
+mezzano.file-system:find-host. This method should:
+
+  * delete the association of the host with the (old) block device;
+  * call the appropriate function for "finding" that file system type
+with the host type and the mount-args;
+  * associate the host with the new block device object returned; and,
+  * set the host field in the read/write object to the host (this
+field does not currently exist in the disk object - see
+[Future Work](#future-work) below).
+
+Local block devices are supported by using the function:
+
+    find-local-block-device (<class name> <partition UUID>) => <block device>
+
+find-local-block-device will call the generic function
+
+    probe-block-device (<class name> <block-device>) =>  <partition UUID> or NIL
+
+on each block device object until a matching partition UUID is found
+and find-local-block-device will return that block
+device. probe-block-device will return NIL if the block device is not
+formatted as the given host type. If no matching block device is
+found, find-local-block-device will return NIL.
+
+probe-block-device is a generic function that specializes on the class
+name argument. Therefore, each file system host class that supports
+block devices, will also need to define a probe-block-device
+method. This method needs to read the header of the block device, make
+enough checks to verify that the device contains a file system of the
+appropriate type, then return the partition UUID (or equivalent).
+
+This approach can be expanded to handle other kinds of file systems by
+creating additional partition functions. For example, for NFS the
+following function might be defined:
+
+    find-nfs-partition (<nfs url>) => <nfs read/write object>
+
+For the currently existing http host, remote host, local host and sys
+host, the mount method does nothing except return success. This is all
+that is required as these hosts do not need to do anything special on
+reboot.
+
+The fat host defines a mount method and an appropriate
+probe-block-device method. And therefore, any fat host saved by a
+snapshot will be available on reboot.
+
+Currently (6 April 2020) the ext4 host mount method returns NIL and no
+probe-block-device method is provided. This means any ext4 host saved
+by a snapshot will **not** be available on reboot.
+
+## Future Work
```

Comment:
Work will need to be done to support hotplug devices. Probably involving uncoupling remounting from booting and having block devices be probed when they are registered.